aeneas/src/ir/VariantSolver.v3

```diff
@@ -83,6 +101,197 @@ def ON_STACK = -1;
+						var bit = u1.!((j >> k) & 1);
+						state[j].scalars[i].bits[longest.start + k] = PackingBit.Fixed(bit);
+					}
+				}
+				explicitTag = (i, Interval(longest.start, longest.start + explicitTagLength));
+				return true;
+			}
+		}
+		return false;
+	}
+	private def canDistinguish(active: Array<bool>) {
+		// TODO - this is where unassigned bits will get assigned
+	}
+	private def checkDistinguishable() -> bool {
+		if (useTagScalar) return true;
+		if (normFields.length <= 1) return true; // single case is always distinguishable
+		var result = tryExplicitTaggingHeuristic();
+		// TODO: try difficult tagging
+		return result;
+	}
+	private def solveField(idx: int) -> bool {
+		if (idx >= fields.length) {
+			return checkDistinguishable();
+		}
+		var field = fields[idx];
+		var scalarAsn = scalarAsns[field];
+
+		// try to squeeze this field into this scalar
+		var casePattern = state[field.caseIdx];
+		var bw = vnorm.nc.GetBitWidth(vnorm.rn.ra.compiler, vnorm.rn.ra.prog, normFields[field.caseIdx][field.fieldIdx]);
+		var scalar = casePattern.scalars[scalarAsn];
+		var intervals = scalar.getIntervalsForSize(bw);
+
+		if (intervals.length == 0) return false;
+		for (interval in intervals) {
+			scalar.assignInterval(interval);
+			intervalAsns[field] = interval;
+			var result = solveField(idx + 1);
+			if (result) return true;
+			scalar.unassignInterval(interval);
+		}
+		return false;
+	}
+	private def getScalarPattern(s: Scalar) -> ScalarPattern {
+		// XXX: Should be based on scalar type
+		return ScalarPattern.new(Array<PackingBit>.new(s.size));
+	}
+	private def solvePacking() -> bool {
+		var cases = Array<VariantPattern>.new(normFields.length);
+		// build up patterns from the individual cases
+		for (i < normFields.length) {
+			var patterns = Array<ScalarPattern>.new(curRep.length);
+			for (j < curRep.length) patterns[j] = getScalarPattern(getScalarFromSet(curRep[j]));
+			cases[i] = VariantPattern.new(patterns);
+		}
+		state = cases;
+
+		return solveField(0);
+	}
+	def render(buf: StringBuilder) -> StringBuilder {
+		for (i < normFields.length) {
+			if (i > 0) buf.csp();
+			buf.put1("c%d: [", i);
+			for (j < normFields[i].length) {
+				if (j > 0) buf.csp();
+				var cf = CaseField(i, j);
+				var field = normFields[i][j];
+				buf.put2("f%d=%q: ", j, field.render);
+
+				if (!scalarAsns.has(cf)) buf.puts("?");
+				else if (!intervalAsns.has(cf)) buf.put1("#%d@?", scalarAsns[cf]);
+				else buf.put2("#%d@%q", scalarAsns[cf], intervalAsns[cf].render);
+			}
+			buf.puts("]");
+		}
+		buf.ln();
+		for (i < state.length) {
+			buf.put1("%q\n", state[i].render);
+		}
+		return buf;
+	}
+	def reset() {
+		scalarAsns = HashMap<CaseField, int>.new(CaseField.hash, CaseField.==);
+		intervalAsns = HashMap<CaseField, Interval>.new(CaseField.hash, CaseField.==);
+		curRep = Vector<Scalar.set>.new();
+		curUsed = Vector<byte>.new();
+		fields = Vector<CaseField>.new();
+	}
+	private def getTagLength(numCases: int) -> byte {
```

Comment:
You should be able to use `Ints.log` here.

Reply:
I think `Ints.log` returns the floor of the log (see `IntsTest.v3`), but I need the ceil here. Is there another method I can use?

Reply:
It's probably worthwhile having a `log` that does the ceiling. Can do in a followup.